File: projects/11-water-treatment-utility/backend/processor/transform.js
"use strict";

// sort_key = window_end#site_id prevents plant-1 and plant-2 readings for
// the same sensor_type in the same flush cycle from colliding on the
// DynamoDB primary key (sensor_type is the partition key, sort_key the
// range key).
function composePlantRangeKey(windowEnd, plantId) {
  return `${windowEnd}#${plantId}`;
}

function windowToReadingItem(rawWindow) {
  const windowData = typeof rawWindow === "string" ? JSON.parse(rawWindow) : rawWindow;
  const siteId = windowData.site_id || "plant-1";
  return {
    sensor_type: windowData.sensor_type,
    sort_key: composePlantRangeKey(windowData.window_end, siteId),
    window_end: windowData.window_end,
    window_start: windowData.window_start,
    site_id: siteId,
    unit: windowData.unit || "",
    count: windowData.count,
    min: windowData.min,
    max: windowData.max,
    avg: windowData.avg,
    latest: windowData.latest,
    alerts: windowData.alerts || [],
  };
}

module.exports = { windowToReadingItem, composePlantRangeKey };

File: projects/17-solar-farm-monitoring/backend/processor/transform.py
import json


def to_item(message_body):
    """Transform a window-aggregate SQS body into the flat DynamoDB record; sort_key is window_end#site_id to keep both arrays distinct."""
    data = json.loads(message_body) if isinstance(message_body, str) else message_body
    site_id = data.get("site_id", "array-1")
    return {
        "sensor_type": data["sensor_type"],
        "sort_key": f"{data['window_end']}#{site_id}",
        "window_start": data["window_start"],
        "window_end": data["window_end"],
        "site_id": site_id,
        "unit": data.get("unit", ""),
        "count": data["count"],
        "min": data["min"],
        "max": data["max"],
        "avg": data["avg"],
        "latest": data["latest"],
        "alerts": data.get("alerts", []),
    }
